### PR TITLE
Allow manual addition of the lines

### PR DESCRIPTION
**Pull request by m-sche**
New button "add line" added. When clicked, it allows the user to place a first end of the new line (denoted by a big red dot). After that, the user can place the second end, which looks just like ordinary moving of the lines.

This is a first draft, any comments are appreciated.

### CODE REVIEW

**Comment by adrian17** [comment]
Didn't have a chance to make a proper review or testing yet, but just a thought... why not just put the line on the main circle (as a diagonal at some angle) and let the user move it himself? I'm not saying it's a better way to do it nor am I forcing it - it feels simpler, but is probably worse from UX perspective.

I'll try your branch tomorrow and see how it looks like.

**Comment by m-sche** [comment]
Yes, that was the first version I did -- add a line with both ends at the main circle and let the user handle it. It felt a bit odd from the user perspective, so I did this more complicated version. Of course, there is this disadvantage of adding complexity to the code, and the need to correctly handle situations like "a user clicks on 'delete line' and than immediately on 'add line'". I am not against the simpler version, I may try to recreate it and post it in a separate pull request, and then we can compare them.

File: gallifreyan.js [hunk 1]
```diff
@@ -19,6 +19,7 @@ var lines           = [],
 var dirtyRender     = true; //whether GUI and red dots will be drawn
 
 var deleteLineMode  = false;//whether next selected line will be deleted
+var addLineMode     = false;//next mouse click will place the first end of a new line
```
Comment (adrian17) [inline gallifreyan.js]: I'd make this comment more explicit:

```
// Both ends of a line move with a cursor. It looks like you're placing the first end of the line.
// A click will disable this mode, and you'll normally control one end of the line.
```

File: gallifreyan.js [hunk 2]
```diff
@@ -88,6 +89,24 @@ function updateText() {
...
+        deleteLine(selectedLine);
+        selectedLine = null;
+    }
+}
+
+//create a new line and let user to position it
+function addNewLine() {
+    addLineMode = true;
+    selectedLine = new Line(allCircles[0], -PI/2, allCircles[0], -PI/2);
+    allCircles[0].lines.removeItem(selectedLine);
```
Comment (adrian17) [inline gallifreyan.js]: I don't think we need this line.

File: gallifreyan.js [hunk 3]
```diff
@@ -255,6 +274,8 @@ function doClick(e) {
         deleteLineMode = false;
         redraw();
     }
+    if (addLineMode)
+        addLineMode = false;	//first end has already been placed, now it's a normal selectedLine
```
Comment (adrian17) [inline gallifreyan.js]: I'd also change the comment here: `don't move both ends anymore; now it's a normal selectedLine with one attached end`

File: gallifreyan.js [hunk 4]
```diff
@@ -344,6 +365,8 @@ $('canvas').mousemove(function(e) {
                 minD = d;
                 a = Math.atan2(mouse.y - allCircles[i].y, mouse.x - allCircles[i].x);
                 selected.updatePoint(lineEnd, allCircles[i], a);
+                if (addLineMode)
+                    selected.updatePoint(1-lineEnd, allCircles[i], a);	//moving both ends at once looks like a single red dot
```
Comment (adrian17) [inline gallifreyan.js]: `(lineEnd+1) % 2` ?

**Comment by adrian17** [comment]
The UX is great.

The code looks okay, it feels a bit like a cheat (well, it is :D), but it's a smart way of making it work with the current system with very little code. It just needs the comments to explain it a bit better, see the notes.

File: gallifreyan.js (hunk 2 above)
Comment (m-sche) [reply gallifreyan.js]: Oh cool, I did't like that line :)

Comment (adrian17) [reply gallifreyan.js]: You left it in the new commit, do you want it to stay after all? (I can merge now)

Comment (m-sche) [reply gallifreyan.js]: It's gone. That was just a mistake. Yes, this commit is more hack-ish, I hope it won't cause any harm, but I can't come up with any better/cleaner solution.

**Pull request merged**